Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -0,0 +1,221 @@
+        {
+            if (_alreadyConnected) throw new ReverseCallClientAlreadyCalledConnect();
+            lock (_connectLock)
+            {
+                if (_alreadyConnected) throw new ReverseCallClientAlreadyCalledConnect();
+                _alreadyConnected = true;
+            }
+
+            var streamingCall = _establishConnection();
+            _clientToServer = streamingCall.RequestStream;
+            _serverToClient = streamingCall.ResponseStream;
+            var callContext = new ReverseCallArgumentsContext
+                {
+                    ExecutionContext = _executionContextManager.Current.ToProtobuf()
+                };
+            _setArgumentsContext(connectArguments, callContext);
+            var message = new TClientMessage();
+            _setConnectArguments(message, connectArguments);
+
+            await _clientToServer.WriteAsync(message).ConfigureAwait(false);
+            if (await _serverToClient.MoveNext(cancellationToken).ConfigureAwait(false))
+            {
+                var response = _getConnectResponse(_serverToClient.Current);
+                if (response != null)
+                {
+                    _logger.Trace("Received connect response");
+                    ConnectResponse = response;
+                    return true;
+                }
+                else
+                {
+                    _logger.Warning("Did not receive connect response. Server message did not contain the connect response");
+                }
+            }
+            else
+            {
+                _logger.Warning("Did not receive connect response. Server stream was empty");
+            }
+
+            await _clientToServer.CompleteAsync().ConfigureAwait(false);
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public async Task Handle(Func<TRequest, CancellationToken, Task<TResponse>> callback, CancellationToken cancellationToken)
+        {
+            if (ConnectResponse == null) throw new ReverseCallClientNotConnected();
+            if (_handleRequests != null)
+            {
+                await _handleRequests.ConfigureAwait(false);
+                return;
+            }
+
+            lock (_handleLock)
+            {
+                if (_handleRequests == null)
+                {
+                    _handleRequests = Task.Run(
+                        async () =>
+                        {
+                            while (await _serverToClient.MoveNext(cancellationToken).ConfigureAwait(false))
+                            {
+                                var request = _getMessageRequest(_serverToClient.Current);
+                                _ = Task.Run(() => HandleRequest(callback, request, cancellationToken));
+                            }
+                        });
+                }
+            }
+
+            await _handleRequests.ConfigureAwait(false);
```

Comment:
This way of doing it indicates that it is OK to call Handle multiple times - and that the first task started will be the one used to do the actual work. I don't think this makes it very clear how this works. And if you pass in different callbacks each time, the other ones will just be ignored.

IMHO we should throw an exception here if it is called multiple times, as I don't see where it would be useful to allow this. If multiple other things needs to wait for the handling to be complete, the task returned the first time can be shared.

Reply:
That is true. If the Handle did not have any arguments I think that this could be a solution. But I see your point

Reply:
I agree.

Reply:
Fixed by https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/43403fe1a3e9cbaf47d70b3e41f57d7ac5f3da9e